### PR TITLE
[refactor] SNode now uses unique_ptr instead of shared_ptr for clearer children ownership

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #518 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: taichi/ir/snode.h [hunk 1]
```diff
@@ -92,14 +96,8 @@ class SNode {
...
 
-  SNode();
-
-  SNode(int depth, SNodeType t);
-
-  ~SNode();
-
   SNode &insert_children(SNodeType t) {
-    ch.push_back(create(depth + 1, t));
+    ch.emplace_back(create(depth + 1, t));
```
Comment (k-ye) [inline taichi/ir/snode.h:100]: Actually `push_back` is probably better, see https://abseil.io/tips/112 (feel like i'm a language lawyer now..)

**Comment by masahi** [comment]
How is this related to 518?

File: taichi/ir/snode.h (hunk 1 above)
Comment (yuanming-hu) [reply taichi/ir/snode.h:100]: Cool! Thanks for pointing this out - glad to learn something new about C++ everyday :-)

**Comment by yuanming-hu** [comment]
> How is this related to 518?

Sorry, I forgot to add a bulleted list regarding this item in #518. This is fixed now.

**Pull request merged**